### PR TITLE
Refine `machine_type` scaling for AnnData files during extract, differential expression (SCP-5769)

### PR DESCRIPTION
**Pull request by bistline**
#### BACKGROUND & CHANGES
This update adds the auto-scaling logic found during AnnData extraction to automated differential expression calculation jobs for AnnData files.  Additionally, the scaling algorithm has been updated such that both parameters classes (`AnnDataIngestParameters` and `DifferentialExpressionParameters`) have their own unique scaling coefficient, meaning they scale differently with respect to file size.  This is controlled via the `GB_PER_CORE` constant in each class.  For extraction jobs, we set a max of 1.75GB/core, and DE jobs have double that (3.5).  This makes the machine types scale as follows:

Extraction machine types
```
n2d-highmem-4: 0-7 GB
n2d-highmem-8: 7-14 GB
n2d-highmem-16: 14-28 GB
n2d-highmem-32: 28-56 GB
n2d-highmem-48: 56-84 GB
n2d-highmem-64: 84-224 GB
```
Differential expression machine types
```
n2d-highmem-4: 0-14 GB
n2d-highmem-8: 14-28 GB
n2d-highmem-16: 28-56 GB
n2d-highmem-32: 56-112 GB
n2d-highmem-48: 112-168 GB
n2d-highmem-64: 168-448 GB
```
Any file larger than the "max" file size is automatically assigned the largest machine type of `n2d-highmem-64`.  Any manual overrides provided are still honored.

#### MANUAL TESTING
1. Enter a Rails console session
2. Create a parameters object for each class using a mock file size of `100.gigabytes`:
```
file_size = 100.gigabytes
extract = AnnDataIngestParameters.new(file_size:)
de = DifferentialExpressionParameters.new(file_size:, matrix_file_type: 'h5ad')
```
3. Confirm the machine types scale to the `32` and `64` core versions, respectively:
```
de.machine_type
=> "n2d-highmem-32"
irb(main):060:0> extract.machine_type
=> "n2d-highmem-64"
```

### CODE REVIEW

File: app/models/differential_expression_parameters.rb [hunk 1]
```diff
@@ -2,10 +2,14 @@
 class DifferentialExpressionParameters
   include ActiveModel::Model
   include Parameterizable
+  include ComputeScaling
 
   # name of Ingest Pipeline CLI parameter that invokes correct parser
   PARAMETER_NAME = '--differential-expression'.freeze
 
+  # scaling coefficient for auto-selecting machine_type
+  GB_PER_CORE = 3.5
```
Comment (eweitz) [inline app/models/differential_expression_parameters.rb:11]: A bit clearer:

```suggestion
  # RAM scaling coefficient for auto-selecting machine_type
  GB_PER_CORE = 3.5
```

File: app/models/ann_data_ingest_parameters.rb [hunk 2]
```diff
@@ -3,6 +3,10 @@
 class AnnDataIngestParameters
   include ActiveModel::Model
   include Parameterizable
+  include ComputeScaling
+
+  # scaling coefficient for auto-selecting machine_type
```
Comment (eweitz) [inline app/models/ann_data_ingest_parameters.rb:8]:
```suggestion
  # RAM scaling coefficient for auto-selecting machine_type
```

File: app/models/concerns/compute_scaling.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# dynamic GCE instance scaling based on input file size
+module ComputeScaling
+  extend ActiveSupport::Concern
+
+  # GCE machine types and file size ranges
+  # produces a hash with entries like { 'n2-highmem-4' => 0..16.gigabytes }
+  # change GB_PER_CORE in parent class to adjust scaling (1-8, lower numbers means more aggressive scaling)
```
Comment (eweitz) [inline app/models/concerns/compute_scaling.rb:7]: Increasing GB_PER_CORE increases memory per core, which seems like _more_ (not less) aggressive scaling, right?

```suggestion
  # change GB_PER_CORE in parent class to adjust scaling (1-8, higher number means more aggressive scaling)
```

**Review by eweitz — APPROVED** [review]
Nice robustness improvement!  Code looks good overall.  I suggest a few comment refinements for clarity and maintainability.

Memory-related errors seem to account for [over 4% of AnnData ingest failures](https://mixpanel.com/s/1buXPB).  It's likely _a lot_ more frequent for visualizable AnnData, which we should get insight on going forward per https://github.com/broadinstitute/single_cell_portal_core/pull/2120.  I haven't found any analytics on DE job failures in Mixpanel, but I suspect memory-related DE ingest failures also have a non-trivial incidence.

Such errors were severe when they occurred for AnnData -- blocking authors, and costing engineer time to manually fix.  

So this update will have good impact!

File: app/models/concerns/compute_scaling.rb (hunk 3 above)
Comment (bistline) [reply app/models/concerns/compute_scaling.rb:7]: Ah yes - you are correct.  Great catch!

**Comment by bistline** [comment]
> Nice robustness improvement! Code looks good overall. I suggest a few comment refinements for clarity and maintainability.
> 
> Memory-related errors seem to account for [over 4% of AnnData ingest failures](https://mixpanel.com/s/1buXPB). It's likely _a lot_ more frequent for visualizable AnnData, which we should get insight on going forward per #2120. I haven't found any analytics on DE job failures in Mixpanel, but I suspect memory-related DE ingest failures also have a non-trivial incidence.
> 
> Such errors were severe when they occurred for AnnData -- blocking authors, and costing engineer time to manually fix.
> 
> So this update will have good impact!

We have this report on automated DE ingest: https://mixpanel.com/project/2120588/view/19059/app/boards#id=3292575&editor-card-id=%22report-34696797%22

**Review by jlchang — APPROVED** [review]
Hoping this will prevent other study owners like Margaret from being blocked by `error 137`. In fact, I'll reach out to Margaret and tell her that re-ingest of any new files should work for her :)

**Pull request merged**